{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ingest gene expression data from TCGA, TARGET and GTEX\n",
    "\n",
    "Download gene expression data from the [UCSC Xena Toil re-compute dataset](https://xenabrowser.net/datapages/?host=https://toil.xenahubs.net), wrangle, enrich and store in an hdf5 file for machine learning. This dataset comprises gene expression data for nineteen thousand samples  processed using the same genomics pipeline and therefore comparable to each other. \n",
    "\n",
    "Each sample consists of a float32 vector, log2(tpm+0.001) normalized, of gene expression for each of 60k genes. Associated with these data is clinical information on each sample such as type (tumor vs. normal), primary site (where the sample came from in the human body) etc... We use this information to label the samples normal/0 vs. tumor/1 as well as to provide additional information for visualization and interpretation of models. \n",
    "\n",
    "The resulting hdf5 file, tumornormal.h5, contains the following datasets:\n",
    "\n",
    "* **X_train, X_test:** float32 gene expression train/test sets split via stratification on class (primary site)\n",
    "* **y_train, y_test:** Binary 0=Normal, 1=Tumor label for binary classification machine learning\n",
    "* **features:** Ensemble gene id's for each feature in X\n",
    "* **labels:** Labels for y binary values ie \"Normal\" = 0 and \"Tumor\" = 1\n",
    "* **classes_train, classes_test:** Class id (integer) for each sample - useful for visualization when clustering\n",
    "* **class_labels:** Text corresponding to each class id (ie disease) - useful as legend when visualizing\n",
    "* **genes:** Hugo gene names corresponding to each feature (may be duplicates) - more recognizable when visualizing inner layers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import requests\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tables  # Required by h5py to write a pandas table\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X: samples=19260 genes=60498\n",
      "CPU times: user 440 ms, sys: 42.5 s, total: 42.9 s\n",
      "Wall time: 42.8 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\"\"\"\n",
    "Download expression data from xena and save in an hdf5 file. This can take around \n",
    "30 minutes between the download and the conversion from tsv into float32 dataframes\n",
    "We download manually vs. passing read_csv a url directly as the latter times\n",
    "out with this size file. Note we invert the expression matrix to conform \n",
    "to machine learning where rows are samples vs. gene expression files where \n",
    "rows are features (gene, transcript, or variant) and columns are \n",
    "instances (sample or single cell)\n",
    "\"\"\"\n",
    "if not os.path.exists(\"data\"):\n",
    "    os.makedirs(\"data\")\n",
    "    \n",
    "if not os.path.exists(\"data/TcgaTargetGtex_rsem_gene_tpm.tsv.gz\"):\n",
    "    print(\"Downloading TCGA, TARGET and GTEX expression data from UCSC Xena\")\n",
    "    r = requests.get(\"https://toil.xenahubs.net/download/TcgaTargetGtex_rsem_gene_tpm.gz\")\n",
    "    r.raise_for_status()\n",
    "    with open(\"data/TcgaTargetGtex_rsem_gene_tpm.tsv.gz\", \"wb\") as f:\n",
    "        for chunk in r.iter_content(32768):\n",
    "            f.write(chunk)\n",
    "\n",
    "if not os.path.exists(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\"):\n",
    "    print(\"Converting expression to dataframe and storing in hdf5 file\")\n",
    "    expression = pd.read_csv(\"data/TcgaTargetGtex_rsem_gene_tpm.tsv.gz\", \n",
    "                             sep=\"\\t\", index_col=0).dropna().astype(np.float32).T\n",
    "    expression.to_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\", mode=\"w\", format=\"fixed\")\n",
    "\n",
    "X = pd.read_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\").sort_index(axis=0)\n",
    "print(\"X: samples={} genes={}\".format(*X.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>ENSG00000242268.2</th>\n",
       "      <th>ENSG00000259041.1</th>\n",
       "      <th>ENSG00000270112.3</th>\n",
       "      <th>ENSG00000167578.16</th>\n",
       "      <th>ENSG00000278814.1</th>\n",
       "      <th>ENSG00000078237.5</th>\n",
       "      <th>ENSG00000269416.5</th>\n",
       "      <th>ENSG00000263642.1</th>\n",
       "      <th>ENSG00000146083.11</th>\n",
       "      <th>ENSG00000158486.13</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000009694.13</th>\n",
       "      <th>ENSG00000238244.3</th>\n",
       "      <th>ENSG00000216352.1</th>\n",
       "      <th>ENSG00000123685.8</th>\n",
       "      <th>ENSG00000267117.1</th>\n",
       "      <th>ENSG00000273233.1</th>\n",
       "      <th>ENSG00000105063.18</th>\n",
       "      <th>ENSG00000231119.2</th>\n",
       "      <th>ENSG00000280861.1</th>\n",
       "      <th>ENSG00000181518.3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.2934</td>\n",
       "      <td>5.1190</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.8488</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>5.1498</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>2.4571</td>\n",
       "      <td>-1.4305</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>2.8178</td>\n",
       "      <td>-0.1031</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>5.1631</td>\n",
       "      <td>-3.3076</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>4.1277</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.6880</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>3.4830</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.9132</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-0.9406</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-1.5105</td>\n",
       "      <td>4.1764</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>4.4067</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.0440</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>4.3841</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>1.5165</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>1.7141</td>\n",
       "      <td>-1.1488</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>4.8768</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <td>-1.2481</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>5.6860</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>1.3679</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>5.0644</td>\n",
       "      <td>-1.8836</td>\n",
       "      <td>...</td>\n",
       "      <td>1.5998</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>3.9356</td>\n",
       "      <td>-1.1488</td>\n",
       "      <td>-1.0559</td>\n",
       "      <td>4.8694</td>\n",
       "      <td>-1.9379</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <td>-3.8160</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.3573</td>\n",
       "      <td>4.0357</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-0.4325</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>3.9421</td>\n",
       "      <td>-3.4580</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.1714</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.6608</td>\n",
       "      <td>-1.1811</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>3.6816</td>\n",
       "      <td>-2.6349</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 60498 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample                    ENSG00000242268.2  ENSG00000259041.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658            -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658            -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS            -1.2481            -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN            -3.8160            -9.9658   \n",
       "\n",
       "sample                    ENSG00000270112.3  ENSG00000167578.16  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -4.2934              5.1190   \n",
       "GTEX-1117F-0426-SM-5EGHI             0.0014              4.1277   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658              4.4067   \n",
       "GTEX-1117F-0626-SM-5N9CS            -5.5735              5.6860   \n",
       "GTEX-1117F-0726-SM-5GIEN             0.3573              4.0357   \n",
       "\n",
       "sample                    ENSG00000278814.1  ENSG00000078237.5  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658             0.8488   \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658             0.6880   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658             0.0440   \n",
       "GTEX-1117F-0626-SM-5N9CS            -9.9658             1.3679   \n",
       "GTEX-1117F-0726-SM-5GIEN            -9.9658            -0.4325   \n",
       "\n",
       "sample                    ENSG00000269416.5  ENSG00000263642.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658            -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658            -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS            -9.9658            -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN            -5.5735            -9.9658   \n",
       "\n",
       "sample                    ENSG00000146083.11  ENSG00000158486.13  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7              5.1498             -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI              3.4830             -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ              4.3841             -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS              5.0644             -1.8836   \n",
       "GTEX-1117F-0726-SM-5GIEN              3.9421             -3.4580   \n",
       "\n",
       "sample                          ...          ENSG00000009694.13  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7        ...                      2.4571   \n",
       "GTEX-1117F-0426-SM-5EGHI        ...                     -0.9132   \n",
       "GTEX-1117F-0526-SM-5EGHJ        ...                      1.5165   \n",
       "GTEX-1117F-0626-SM-5N9CS        ...                      1.5998   \n",
       "GTEX-1117F-0726-SM-5GIEN        ...                     -3.1714   \n",
       "\n",
       "sample                    ENSG00000238244.3  ENSG00000216352.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -1.4305            -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658            -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS            -9.9658            -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN            -9.9658            -9.9658   \n",
       "\n",
       "sample                    ENSG00000123685.8  ENSG00000267117.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7             2.8178            -0.1031   \n",
       "GTEX-1117F-0426-SM-5EGHI            -0.9406            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ             1.7141            -1.1488   \n",
       "GTEX-1117F-0626-SM-5N9CS             3.9356            -1.1488   \n",
       "GTEX-1117F-0726-SM-5GIEN             0.6608            -1.1811   \n",
       "\n",
       "sample                    ENSG00000273233.1  ENSG00000105063.18  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658              5.1631   \n",
       "GTEX-1117F-0426-SM-5EGHI            -1.5105              4.1764   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658              4.8768   \n",
       "GTEX-1117F-0626-SM-5N9CS            -1.0559              4.8694   \n",
       "GTEX-1117F-0726-SM-5GIEN            -9.9658              3.6816   \n",
       "\n",
       "sample                    ENSG00000231119.2  ENSG00000280861.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -3.3076            -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI            -5.0116            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658            -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS            -1.9379            -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN            -2.6349            -9.9658   \n",
       "\n",
       "sample                    ENSG00000181518.3  \n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658  \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658  \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658  \n",
       "GTEX-1117F-0626-SM-5N9CS            -9.9658  \n",
       "GTEX-1117F-0726-SM-5GIEN            -9.9658  \n",
       "\n",
       "[5 rows x 60498 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        category                 disease  \\\n",
       "id                                                                         \n",
       "GTEX-1117F-0226-SM-5GZZ7  Adipose - Subcutaneous  Adipose - Subcutaneous   \n",
       "\n",
       "                            primary_site    sample_type  gender study  \\\n",
       "id                                                                      \n",
       "GTEX-1117F-0226-SM-5GZZ7  Adipose Tissue  Normal Tissue  Female  GTEX   \n",
       "\n",
       "                         tumor_normal  \n",
       "id                                     \n",
       "GTEX-1117F-0226-SM-5GZZ7       Normal  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the sample labels from Xena ie clinical/phenotype information on each sample\n",
    "Y = pd.read_table(\"https://toil.xenahubs.net/download/TcgaTargetGTEX_phenotype.txt.gz\", compression=\"gzip\",\n",
    "                  header=0, names=[\"id\", \"category\", \"disease\", \"primary_site\", \"sample_type\", \"gender\", \"study\"],\n",
    "                  sep=\"\\t\", encoding=\"ISO-8859-1\", index_col=0, dtype=\"str\").sort_index(axis=0)\n",
    "\n",
    "# Compute and add a tumor/normal column - TCGA and TARGET have some normal samples, GTEX is all normal.\n",
    "Y[\"tumor_normal\"] = Y.apply(\n",
    "    lambda row: \"Normal\" if row[\"sample_type\"] in [\"Cell Line\", \"Normal Tissue\", \"Solid Tissue Normal\"]\n",
    "    else \"Tumor\", axis=1)\n",
    "\n",
    "Y[0:100:4000].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>19130</td>\n",
       "      <td>19130</td>\n",
       "      <td>19126</td>\n",
       "      <td>19131</td>\n",
       "      <td>18972</td>\n",
       "      <td>19131</td>\n",
       "      <td>19131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>46</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Male</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1212</td>\n",
       "      <td>1212</td>\n",
       "      <td>1846</td>\n",
       "      <td>9185</td>\n",
       "      <td>10456</td>\n",
       "      <td>10535</td>\n",
       "      <td>10531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         category                    disease primary_site  \\\n",
       "count                       19130                      19130        19126   \n",
       "unique                         93                         93           46   \n",
       "top     Breast Invasive Carcinoma  Breast Invasive Carcinoma        Brain   \n",
       "freq                         1212                       1212         1846   \n",
       "\n",
       "          sample_type gender  study tumor_normal  \n",
       "count           19131  18972  19131        19131  \n",
       "unique             17      2      3            2  \n",
       "top     Primary Tumor   Male   TCGA        Tumor  \n",
       "freq             9185  10456  10535        10531  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the tissue location as the class label for the purposes of stratification\n",
    "class_attribute = \"primary_site\"\n",
    "\n",
    "# Tumor vs. Normal is the binary attribute we'll use to train on\n",
    "label_attribute = \"tumor_normal\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19126 samples with non-null labels\n"
     ]
    }
   ],
   "source": [
    "# Remove rows where the class is null or the sample is missing\n",
    "Y_not_null = Y[pd.notnull(Y[class_attribute])]\n",
    "intersection = X.index.intersection(Y_not_null.index)\n",
    "X_clean = X[X.index.isin(intersection)]\n",
    "Y_clean = Y[Y.index.isin(intersection)]\n",
    "\n",
    "# Make sure the label and example samples are in the same order\n",
    "assert(X_clean.index.equals(Y_clean.index))\n",
    "\n",
    "print(intersection.shape[0], \"samples with non-null labels\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert tumor/normal labels to binary 1/0\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "encoder = LabelEncoder()\n",
    "y_binary = encoder.fit_transform(Y_clean[\"tumor_normal\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total classes for stratification: 46\n"
     ]
    }
   ],
   "source": [
    "# Convert classes into numbers\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "encoder = LabelEncoder()\n",
    "encoder.fit(Y_clean[class_attribute].values)\n",
    "classes = encoder.transform(Y_clean[class_attribute])\n",
    "print(\"Total classes for stratification:\", len(set(classes)))\n",
    "class_labels = encoder.classes_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.96 s, sys: 1.4 s, total: 3.36 s\n",
      "Wall time: 3.36 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Split into stratified training and test sets based on classes (i.e. tissue type) so that we have equal\n",
    "# proportions of each tissue type in the train and test sets\n",
    "from sklearn.model_selection import StratifiedShuffleSplit\n",
    "split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)\n",
    "for train_index, test_index in split.split(X_clean.values, Y_clean[class_attribute]):\n",
    "    X_train, X_test = X.values[train_index], X_clean.values[test_index]\n",
    "    y_train, y_test = y_binary[train_index], y_binary[test_index]\n",
    "    classes_train, classes_test = classes[train_index], classes[test_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneType</th>\n",
       "      <th>geneStatus</th>\n",
       "      <th>transcriptId</th>\n",
       "      <th>transcriptName</th>\n",
       "      <th>transcriptType</th>\n",
       "      <th>transcriptStatus</th>\n",
       "      <th>havanaGeneId</th>\n",
       "      <th>havanaTranscriptId</th>\n",
       "      <th>ccdsId</th>\n",
       "      <th>level</th>\n",
       "      <th>transcriptClass</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>DDX11L1-001</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>WASH7P-001</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>nonCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.3</th>\n",
       "      <td>RP11-34P13.3</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>RP11-34P13.3-001</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>nonCoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       geneName                            geneType  \\\n",
       "geneId                                                                \n",
       "ENSG00000223972.5       DDX11L1  transcribed_unprocessed_pseudogene   \n",
       "ENSG00000223972.5       DDX11L1  transcribed_unprocessed_pseudogene   \n",
       "ENSG00000227232.5        WASH7P              unprocessed_pseudogene   \n",
       "ENSG00000278267.1     MIR6859-1                               miRNA   \n",
       "ENSG00000243485.3  RP11-34P13.3                             lincRNA   \n",
       "\n",
       "                  geneStatus       transcriptId    transcriptName  \\\n",
       "geneId                                                              \n",
       "ENSG00000223972.5      KNOWN  ENST00000456328.2       DDX11L1-002   \n",
       "ENSG00000223972.5      KNOWN  ENST00000450305.2       DDX11L1-001   \n",
       "ENSG00000227232.5      KNOWN  ENST00000488147.1        WASH7P-001   \n",
       "ENSG00000278267.1      KNOWN  ENST00000619216.1     MIR6859-1-201   \n",
       "ENSG00000243485.3      KNOWN  ENST00000473358.1  RP11-34P13.3-001   \n",
       "\n",
       "                                       transcriptType transcriptStatus  \\\n",
       "geneId                                                                   \n",
       "ENSG00000223972.5                processed_transcript            KNOWN   \n",
       "ENSG00000223972.5  transcribed_unprocessed_pseudogene            KNOWN   \n",
       "ENSG00000227232.5              unprocessed_pseudogene            KNOWN   \n",
       "ENSG00000278267.1                               miRNA            KNOWN   \n",
       "ENSG00000243485.3                             lincRNA            KNOWN   \n",
       "\n",
       "                           havanaGeneId    havanaTranscriptId ccdsId  level  \\\n",
       "geneId                                                                        \n",
       "ENSG00000223972.5  OTTHUMG00000000961.2  OTTHUMT00000362751.1    NaN      2   \n",
       "ENSG00000223972.5  OTTHUMG00000000961.2  OTTHUMT00000002844.2    NaN      2   \n",
       "ENSG00000227232.5  OTTHUMG00000000958.1  OTTHUMT00000002839.1    NaN      2   \n",
       "ENSG00000278267.1                   NaN                   NaN    NaN      3   \n",
       "ENSG00000243485.3  OTTHUMG00000000959.2  OTTHUMT00000002840.1    NaN      2   \n",
       "\n",
       "                  transcriptClass  \n",
       "geneId                             \n",
       "ENSG00000223972.5          pseudo  \n",
       "ENSG00000223972.5          pseudo  \n",
       "ENSG00000227232.5          pseudo  \n",
       "ENSG00000278267.1       nonCoding  \n",
       "ENSG00000243485.3       nonCoding  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Feature labels are ensemble ids, convert to hugo gene names for use in interpreting\n",
    "hidden layers in any trained models as they are better known to most bioinformaticians \n",
    "and clinicians. We're using an assembled table from John Vivian @ UCSC here. Another\n",
    "option would be ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/hgnc_complete_set.txt\n",
    "\"\"\"\n",
    "ensemble_to_hugo = pd.read_table(\n",
    "    \"https://github.com/jvivian/docker_tools/blob/master/gencode_hugo_mapping/attrs.tsv?raw=true\",\n",
    "    index_col=0)\n",
    "ensemble_to_hugo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The ensembl to hugo table has duplicates due to the many transcripts that map\n",
    "# to a gene. Remove the duplicates and then lookup hugo for each ensemble id\n",
    "hugo = ensemble_to_hugo[~ensemble_to_hugo.index.duplicated(keep='first')].reindex(X_clean.columns.values)[\"geneName\"].fillna(\"\")\n",
    "\n",
    "# Make sure we end up with the order of features being identical as some ensemble id's\n",
    "# map to the same hugo gene id\n",
    "assert(X_clean.columns.equals(hugo.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Write to an h5 file for training (see above for details on each dataset)\n",
    "\"\"\"\n",
    "with h5py.File(\"data/tumor_normal.h5\", \"w\") as f:\n",
    "    f.create_dataset('X_train', X_train.shape, dtype='f')[:] = X_train\n",
    "    f.create_dataset('X_test', X_test.shape, dtype='f')[:] = X_test\n",
    "    f.create_dataset('y_train', y_train.shape, dtype='i')[:] = y_train\n",
    "    f.create_dataset('y_test', y_test.shape, dtype='i')[:] = y_test\n",
    "    f.create_dataset('classes_train', y_train.shape, dtype='i')[:] = classes_train\n",
    "    f.create_dataset('classes_test', y_test.shape, dtype='i')[:] = classes_test\n",
    "    f.create_dataset('features', X_clean.columns.shape, 'S10', \n",
    "                     [l.encode(\"ascii\", \"ignore\") for l in X_clean.columns.values])\n",
    "    f.create_dataset('genes', hugo.shape, 'S10', \n",
    "                     [l.encode(\"ascii\", \"ignore\") for l in hugo.values.tolist()])\n",
    "    f.create_dataset('labels', (2, 1), 'S10', \n",
    "                     [l.encode(\"ascii\", \"ignore\") for l in [\"Normal\", \"Tumor\"]])\n",
    "    f.create_dataset('class_labels', (len(class_labels), 1), 'S10', \n",
    "                     [l.encode(\"ascii\", \"ignore\") for l in class_labels])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
