{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ingest Gene Expression and Clinical Data from TCGA+TARGET+GTEX and Treehouse\n",
    "\n",
    "Download gene expression and clinical data from the [UCSC Xena Toil re-compute dataset](https://xenabrowser.net/datapages/?host=https://toil.xenahubs.net) and the [Treehouse Childhood Cancer Initiative](https://xenabrowser.net/datapages/?host=https://treehouse.xenahubs.net), wrangle, and store in an hdf5 file for quick loading machine learning. This dataset comprises gene expression data for twenty thousand tumor and normal samples processed using the exact same genomics pipeline and therefore can be compared to each other. Treehouse contains many of the same samples from TCGA and TARGET as Toil which we can use to verify our conversion. It also includes unique samples (all prefixed with TH or TR) which we can use as a hold-out set.\n",
    "\n",
    "Each of the source data set consists of a float vector, log2(TPM+0.001) in the case of TCGA+TARGET+GTEX or log2(TPM+1.0) in the case of Treehouse normalized, of gene expression for each of ~60k genes. Toil expression is labeled using Ensembl gene ids vs. Treehouse which uses Hugo. Associated with these data is clinical information on each sample such as type (tumor vs. normal), disease, primary site (where the sample came from in the human body) etc... We use this information to label the samples normal/0 vs. tumor/1 as well as to provide additional information for visualization and interpretation of models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import requests\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import h5py\n",
    "\n",
    "if not os.path.exists(\"data\"):\n",
    "    os.makedirs(\"data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download TGCA+TARGET+GTEX Expression\n",
    "Download expression data files from Xena and save in an hdf5 file. This can take around 30 minutes each between the download and the conversion from tsv into float32 dataframes. We download manually vs. passing read_csv a url directly as the latter times out with this size file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tcga_target_gtex_expression: samples=19260 genes=60498\n",
      "CPU times: user 6.74 s, sys: 7.26 s, total: 14 s\n",
      "Wall time: 14 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if not os.path.exists(\"data/TcgaTargetGtex_rsem_gene_tpm.gz\"):\n",
    "    print(\"Downloading TCGA, TARGET and GTEX expression data from UCSC Xena\")\n",
    "    r = requests.get(\"https://toil.xenahubs.net/download/TcgaTargetGtex_rsem_gene_tpm.gz\", stream=True)\n",
    "    response.raise_for_status()\n",
    "    with open(\"data/TcgaTargetGtex_rsem_gene_tpm.gz\", \"wb\") as f:\n",
    "        for chunk in r.iter_content(chunk_size=32768):\n",
    "            f.write(chunk)\n",
    "\n",
    "if not os.path.exists(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\"):\n",
    "    print(\"Converting expression to dataframe and storing in hdf5 file\")\n",
    "    pd.read_csv(\"data/TcgaTargetGtex_rsem_gene_tpm.gz\", sep=\"\\t\", index_col=0) \\\n",
    "        .astype(np.float32).to_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\", mode=\"w\", format=\"fixed\")\n",
    "\n",
    "tcga_target_gtex_expression = pd.read_hdf(\n",
    "    \"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\").T.dropna(axis=\"index\").sort_index(axis=\"columns\")\n",
    "print(\"tcga_target_gtex_expression: samples={} genes={}\".format(*tcga_target_gtex_expression.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <th>ENSG00000000938.12</th>\n",
       "      <th>ENSG00000000971.15</th>\n",
       "      <th>ENSG00000001036.13</th>\n",
       "      <th>ENSG00000001084.10</th>\n",
       "      <th>ENSG00000001167.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSGR0000263980.5</th>\n",
       "      <th>ENSGR0000264510.5</th>\n",
       "      <th>ENSGR0000264819.5</th>\n",
       "      <th>ENSGR0000265658.5</th>\n",
       "      <th>ENSGR0000270726.5</th>\n",
       "      <th>ENSGR0000275287.4</th>\n",
       "      <th>ENSGR0000276543.4</th>\n",
       "      <th>ENSGR0000277120.4</th>\n",
       "      <th>ENSGR0000280767.2</th>\n",
       "      <th>ENSGR0000281849.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-UTHO-1226-SM-3GAEE</th>\n",
       "      <td>3.0056</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>4.5098</td>\n",
       "      <td>1.7403</td>\n",
       "      <td>1.3340</td>\n",
       "      <td>4.1883</td>\n",
       "      <td>8.9373</td>\n",
       "      <td>4.0994</td>\n",
       "      <td>2.9581</td>\n",
       "      <td>3.2526</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-146FH-1726-SM-5QGQ2</th>\n",
       "      <td>5.9729</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>5.0922</td>\n",
       "      <td>3.1572</td>\n",
       "      <td>2.3308</td>\n",
       "      <td>2.0569</td>\n",
       "      <td>4.5632</td>\n",
       "      <td>3.5742</td>\n",
       "      <td>4.4344</td>\n",
       "      <td>3.3633</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-QDT8-0126-SM-48TZ1</th>\n",
       "      <td>3.6939</td>\n",
       "      <td>2.4675</td>\n",
       "      <td>4.9810</td>\n",
       "      <td>2.7972</td>\n",
       "      <td>1.4808</td>\n",
       "      <td>3.9792</td>\n",
       "      <td>6.8849</td>\n",
       "      <td>4.2419</td>\n",
       "      <td>3.6417</td>\n",
       "      <td>3.4411</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-QCQG-1326-SM-48U24</th>\n",
       "      <td>4.9594</td>\n",
       "      <td>-3.4580</td>\n",
       "      <td>5.2223</td>\n",
       "      <td>2.7357</td>\n",
       "      <td>1.8564</td>\n",
       "      <td>3.7061</td>\n",
       "      <td>2.0218</td>\n",
       "      <td>5.1559</td>\n",
       "      <td>4.3449</td>\n",
       "      <td>3.5535</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-WZTO-2926-SM-3NM9I</th>\n",
       "      <td>2.8760</td>\n",
       "      <td>-2.4659</td>\n",
       "      <td>4.7022</td>\n",
       "      <td>1.9034</td>\n",
       "      <td>0.0440</td>\n",
       "      <td>1.6558</td>\n",
       "      <td>3.1733</td>\n",
       "      <td>3.1062</td>\n",
       "      <td>3.6657</td>\n",
       "      <td>2.8482</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 60498 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample                    ENSG00000000003.14  ENSG00000000005.5  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               3.0056            -9.9658   \n",
       "GTEX-146FH-1726-SM-5QGQ2              5.9729            -9.9658   \n",
       "GTEX-QDT8-0126-SM-48TZ1               3.6939             2.4675   \n",
       "GTEX-QCQG-1326-SM-48U24               4.9594            -3.4580   \n",
       "GTEX-WZTO-2926-SM-3NM9I               2.8760            -2.4659   \n",
       "\n",
       "sample                    ENSG00000000419.12  ENSG00000000457.13  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               4.5098              1.7403   \n",
       "GTEX-146FH-1726-SM-5QGQ2              5.0922              3.1572   \n",
       "GTEX-QDT8-0126-SM-48TZ1               4.9810              2.7972   \n",
       "GTEX-QCQG-1326-SM-48U24               5.2223              2.7357   \n",
       "GTEX-WZTO-2926-SM-3NM9I               4.7022              1.9034   \n",
       "\n",
       "sample                    ENSG00000000460.16  ENSG00000000938.12  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               1.3340              4.1883   \n",
       "GTEX-146FH-1726-SM-5QGQ2              2.3308              2.0569   \n",
       "GTEX-QDT8-0126-SM-48TZ1               1.4808              3.9792   \n",
       "GTEX-QCQG-1326-SM-48U24               1.8564              3.7061   \n",
       "GTEX-WZTO-2926-SM-3NM9I               0.0440              1.6558   \n",
       "\n",
       "sample                    ENSG00000000971.15  ENSG00000001036.13  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               8.9373              4.0994   \n",
       "GTEX-146FH-1726-SM-5QGQ2              4.5632              3.5742   \n",
       "GTEX-QDT8-0126-SM-48TZ1               6.8849              4.2419   \n",
       "GTEX-QCQG-1326-SM-48U24               2.0218              5.1559   \n",
       "GTEX-WZTO-2926-SM-3NM9I               3.1733              3.1062   \n",
       "\n",
       "sample                    ENSG00000001084.10  ENSG00000001167.14  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE               2.9581              3.2526   \n",
       "GTEX-146FH-1726-SM-5QGQ2              4.4344              3.3633   \n",
       "GTEX-QDT8-0126-SM-48TZ1               3.6417              3.4411   \n",
       "GTEX-QCQG-1326-SM-48U24               4.3449              3.5535   \n",
       "GTEX-WZTO-2926-SM-3NM9I               3.6657              2.8482   \n",
       "\n",
       "sample                          ...          ENSGR0000263980.5  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE         ...                    -9.9658   \n",
       "GTEX-146FH-1726-SM-5QGQ2        ...                    -9.9658   \n",
       "GTEX-QDT8-0126-SM-48TZ1         ...                    -9.9658   \n",
       "GTEX-QCQG-1326-SM-48U24         ...                    -9.9658   \n",
       "GTEX-WZTO-2926-SM-3NM9I         ...                    -9.9658   \n",
       "\n",
       "sample                    ENSGR0000264510.5  ENSGR0000264819.5  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE             -9.9658            -9.9658   \n",
       "GTEX-146FH-1726-SM-5QGQ2            -9.9658            -9.9658   \n",
       "GTEX-QDT8-0126-SM-48TZ1             -9.9658            -9.9658   \n",
       "GTEX-QCQG-1326-SM-48U24             -9.9658            -9.9658   \n",
       "GTEX-WZTO-2926-SM-3NM9I             -9.9658            -9.9658   \n",
       "\n",
       "sample                    ENSGR0000265658.5  ENSGR0000270726.5  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE             -9.9658            -9.9658   \n",
       "GTEX-146FH-1726-SM-5QGQ2            -9.9658            -9.9658   \n",
       "GTEX-QDT8-0126-SM-48TZ1             -9.9658            -9.9658   \n",
       "GTEX-QCQG-1326-SM-48U24             -9.9658            -9.9658   \n",
       "GTEX-WZTO-2926-SM-3NM9I             -9.9658            -9.9658   \n",
       "\n",
       "sample                    ENSGR0000275287.4  ENSGR0000276543.4  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE             -9.9658            -9.9658   \n",
       "GTEX-146FH-1726-SM-5QGQ2            -9.9658            -9.9658   \n",
       "GTEX-QDT8-0126-SM-48TZ1             -9.9658            -9.9658   \n",
       "GTEX-QCQG-1326-SM-48U24             -9.9658            -9.9658   \n",
       "GTEX-WZTO-2926-SM-3NM9I             -9.9658            -9.9658   \n",
       "\n",
       "sample                    ENSGR0000277120.4  ENSGR0000280767.2  \\\n",
       "GTEX-UTHO-1226-SM-3GAEE             -9.9658            -9.9658   \n",
       "GTEX-146FH-1726-SM-5QGQ2            -9.9658            -9.9658   \n",
       "GTEX-QDT8-0126-SM-48TZ1             -9.9658            -9.9658   \n",
       "GTEX-QCQG-1326-SM-48U24             -9.9658            -9.9658   \n",
       "GTEX-WZTO-2926-SM-3NM9I             -9.9658            -9.9658   \n",
       "\n",
       "sample                    ENSGR0000281849.2  \n",
       "GTEX-UTHO-1226-SM-3GAEE             -9.9658  \n",
       "GTEX-146FH-1726-SM-5QGQ2            -9.9658  \n",
       "GTEX-QDT8-0126-SM-48TZ1             -9.9658  \n",
       "GTEX-QCQG-1326-SM-48U24             -9.9658  \n",
       "GTEX-WZTO-2926-SM-3NM9I             -9.9658  \n",
       "\n",
       "[5 rows x 60498 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_target_gtex_expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Covert Ensembl to Hugo\n",
    "\n",
    "Toil's expression values are per Ensembl gene id, which have a one or more to one relationship to Hugo gene names so we need to convert back into TPM, average (or add?), and then convert back to log2(tpm+0.001). We're using an assembled table from John Vivian @ UCSC here. Another\n",
    "option would be ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/hgnc_complete_set.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneType</th>\n",
       "      <th>geneStatus</th>\n",
       "      <th>transcriptId</th>\n",
       "      <th>transcriptName</th>\n",
       "      <th>transcriptType</th>\n",
       "      <th>transcriptStatus</th>\n",
       "      <th>havanaGeneId</th>\n",
       "      <th>havanaTranscriptId</th>\n",
       "      <th>ccdsId</th>\n",
       "      <th>level</th>\n",
       "      <th>transcriptClass</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000612152.4</td>\n",
       "      <td>TSPAN6-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000022002.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS76001.1</td>\n",
       "      <td>3</td>\n",
       "      <td>coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000373031.4</td>\n",
       "      <td>TNMD-001</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000022001.1</td>\n",
       "      <td>OTTHUMT00000057481.1</td>\n",
       "      <td>CCDS14469.1</td>\n",
       "      <td>2</td>\n",
       "      <td>coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000371582.8</td>\n",
       "      <td>DPM1-005</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000032742.2</td>\n",
       "      <td>OTTHUMT00000079720.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000470238.1</td>\n",
       "      <td>SCYL3-004</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000035941.4</td>\n",
       "      <td>OTTHUMT00000087552.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>nonCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000466580.6</td>\n",
       "      <td>C1orf112-008</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000035821.7</td>\n",
       "      <td>OTTHUMT00000087524.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>nonCoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    geneName        geneType geneStatus       transcriptId  \\\n",
       "geneId                                                                       \n",
       "ENSG00000000003.14    TSPAN6  protein_coding      KNOWN  ENST00000612152.4   \n",
       "ENSG00000000005.5       TNMD  protein_coding      KNOWN  ENST00000373031.4   \n",
       "ENSG00000000419.12      DPM1  protein_coding      KNOWN  ENST00000371582.8   \n",
       "ENSG00000000457.13     SCYL3  protein_coding      KNOWN  ENST00000470238.1   \n",
       "ENSG00000000460.16  C1orf112  protein_coding      KNOWN  ENST00000466580.6   \n",
       "\n",
       "                   transcriptName        transcriptType transcriptStatus  \\\n",
       "geneId                                                                     \n",
       "ENSG00000000003.14     TSPAN6-201        protein_coding            KNOWN   \n",
       "ENSG00000000005.5        TNMD-001        protein_coding            KNOWN   \n",
       "ENSG00000000419.12       DPM1-005        protein_coding            KNOWN   \n",
       "ENSG00000000457.13      SCYL3-004  processed_transcript            KNOWN   \n",
       "ENSG00000000460.16   C1orf112-008  processed_transcript            KNOWN   \n",
       "\n",
       "                            havanaGeneId    havanaTranscriptId       ccdsId  \\\n",
       "geneId                                                                        \n",
       "ENSG00000000003.14  OTTHUMG00000022002.1                   NaN  CCDS76001.1   \n",
       "ENSG00000000005.5   OTTHUMG00000022001.1  OTTHUMT00000057481.1  CCDS14469.1   \n",
       "ENSG00000000419.12  OTTHUMG00000032742.2  OTTHUMT00000079720.2          NaN   \n",
       "ENSG00000000457.13  OTTHUMG00000035941.4  OTTHUMT00000087552.1          NaN   \n",
       "ENSG00000000460.16  OTTHUMG00000035821.7  OTTHUMT00000087524.1          NaN   \n",
       "\n",
       "                    level transcriptClass  \n",
       "geneId                                     \n",
       "ENSG00000000003.14      3          coding  \n",
       "ENSG00000000005.5       2          coding  \n",
       "ENSG00000000419.12      2          coding  \n",
       "ENSG00000000457.13      2       nonCoding  \n",
       "ENSG00000000460.16      2       nonCoding  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if not os.path.exists(\"data/ensembl_to_hugo.tsv\"):\n",
    "    with open(\"data/ensembl_to_hugo.tsv\", \"wb\") as f:\n",
    "        f.write(requests.get(\"https://github.com/jvivian/docker_tools/blob/master/gencode_hugo_mapping/attrs.tsv?raw=true\").content)\n",
    "ensemble_to_hugo = pd.read_table(\"data/ensembl_to_hugo.tsv\",index_col=0).sort_index(axis=\"index\")\n",
    "\n",
    "# Remove duplicates\n",
    "ensemble_to_hugo = ensemble_to_hugo[~ensemble_to_hugo.index.duplicated(keep='first')]\n",
    "ensemble_to_hugo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-UTHO-1226-SM-3GAEE</th>\n",
       "      <th>GTEX-146FH-1726-SM-5QGQ2</th>\n",
       "      <th>GTEX-QDT8-0126-SM-48TZ1</th>\n",
       "      <th>GTEX-QCQG-1326-SM-48U24</th>\n",
       "      <th>GTEX-WZTO-2926-SM-3NM9I</th>\n",
       "      <th>GTEX-12WSB-0126-SM-59HJN</th>\n",
       "      <th>GTEX-11VI4-0626-SM-5EQLO</th>\n",
       "      <th>GTEX-T5JC-0526-SM-32PM7</th>\n",
       "      <th>GTEX-RU1J-0426-SM-46MUK</th>\n",
       "      <th>GTEX-1212Z-0226-SM-59HLF</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-AB-2965-03</th>\n",
       "      <th>TCGA-AB-2936-03</th>\n",
       "      <th>TCGA-AB-2839-03</th>\n",
       "      <th>TCGA-AB-2879-03</th>\n",
       "      <th>TCGA-AB-2886-03</th>\n",
       "      <th>TCGA-AB-2901-03</th>\n",
       "      <th>TCGA-AB-2862-03</th>\n",
       "      <th>TCGA-AB-2956-03</th>\n",
       "      <th>TCGA-AB-2987-03</th>\n",
       "      <th>TCGA-AB-2868-03</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>3.0056</td>\n",
       "      <td>5.9729</td>\n",
       "      <td>3.6939</td>\n",
       "      <td>4.9594</td>\n",
       "      <td>2.8760</td>\n",
       "      <td>3.8360</td>\n",
       "      <td>3.1028</td>\n",
       "      <td>4.2571</td>\n",
       "      <td>2.3816</td>\n",
       "      <td>4.5838</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.2828</td>\n",
       "      <td>-3.3076</td>\n",
       "      <td>-3.1714</td>\n",
       "      <td>-0.8084</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-0.1828</td>\n",
       "      <td>-4.2934</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>-1.2481</td>\n",
       "      <td>-0.9132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>2.4675</td>\n",
       "      <td>-3.4580</td>\n",
       "      <td>-2.4659</td>\n",
       "      <td>0.6608</td>\n",
       "      <td>-2.8262</td>\n",
       "      <td>3.5073</td>\n",
       "      <td>-2.4659</td>\n",
       "      <td>-2.9324</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.4580</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>4.5098</td>\n",
       "      <td>5.0922</td>\n",
       "      <td>4.9810</td>\n",
       "      <td>5.2223</td>\n",
       "      <td>4.7022</td>\n",
       "      <td>5.2250</td>\n",
       "      <td>4.7814</td>\n",
       "      <td>4.6206</td>\n",
       "      <td>5.1322</td>\n",
       "      <td>4.0471</td>\n",
       "      <td>...</td>\n",
       "      <td>5.2510</td>\n",
       "      <td>4.6566</td>\n",
       "      <td>4.7603</td>\n",
       "      <td>4.4156</td>\n",
       "      <td>4.8719</td>\n",
       "      <td>4.8625</td>\n",
       "      <td>4.5361</td>\n",
       "      <td>5.3345</td>\n",
       "      <td>5.1223</td>\n",
       "      <td>4.1285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>1.7403</td>\n",
       "      <td>3.1572</td>\n",
       "      <td>2.7972</td>\n",
       "      <td>2.7357</td>\n",
       "      <td>1.9034</td>\n",
       "      <td>3.5851</td>\n",
       "      <td>2.2144</td>\n",
       "      <td>1.7912</td>\n",
       "      <td>2.4753</td>\n",
       "      <td>1.7912</td>\n",
       "      <td>...</td>\n",
       "      <td>2.6670</td>\n",
       "      <td>3.9147</td>\n",
       "      <td>3.3435</td>\n",
       "      <td>3.5535</td>\n",
       "      <td>3.5249</td>\n",
       "      <td>3.5060</td>\n",
       "      <td>2.1509</td>\n",
       "      <td>2.9929</td>\n",
       "      <td>2.5011</td>\n",
       "      <td>2.9507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>1.3340</td>\n",
       "      <td>2.3308</td>\n",
       "      <td>1.4808</td>\n",
       "      <td>1.8564</td>\n",
       "      <td>0.0440</td>\n",
       "      <td>2.1247</td>\n",
       "      <td>0.6608</td>\n",
       "      <td>1.0363</td>\n",
       "      <td>1.3109</td>\n",
       "      <td>0.1776</td>\n",
       "      <td>...</td>\n",
       "      <td>3.1939</td>\n",
       "      <td>4.3896</td>\n",
       "      <td>3.5174</td>\n",
       "      <td>3.7105</td>\n",
       "      <td>4.8954</td>\n",
       "      <td>4.5681</td>\n",
       "      <td>2.6918</td>\n",
       "      <td>4.0304</td>\n",
       "      <td>2.4985</td>\n",
       "      <td>3.1491</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19260 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GTEX-UTHO-1226-SM-3GAEE  GTEX-146FH-1726-SM-5QGQ2  \\\n",
       "TSPAN6                     3.0056                    5.9729   \n",
       "TNMD                      -9.9658                   -9.9658   \n",
       "DPM1                       4.5098                    5.0922   \n",
       "SCYL3                      1.7403                    3.1572   \n",
       "C1orf112                   1.3340                    2.3308   \n",
       "\n",
       "          GTEX-QDT8-0126-SM-48TZ1  GTEX-QCQG-1326-SM-48U24  \\\n",
       "TSPAN6                     3.6939                   4.9594   \n",
       "TNMD                       2.4675                  -3.4580   \n",
       "DPM1                       4.9810                   5.2223   \n",
       "SCYL3                      2.7972                   2.7357   \n",
       "C1orf112                   1.4808                   1.8564   \n",
       "\n",
       "          GTEX-WZTO-2926-SM-3NM9I  GTEX-12WSB-0126-SM-59HJN  \\\n",
       "TSPAN6                     2.8760                    3.8360   \n",
       "TNMD                      -2.4659                    0.6608   \n",
       "DPM1                       4.7022                    5.2250   \n",
       "SCYL3                      1.9034                    3.5851   \n",
       "C1orf112                   0.0440                    2.1247   \n",
       "\n",
       "          GTEX-11VI4-0626-SM-5EQLO  GTEX-T5JC-0526-SM-32PM7  \\\n",
       "TSPAN6                      3.1028                   4.2571   \n",
       "TNMD                       -2.8262                   3.5073   \n",
       "DPM1                        4.7814                   4.6206   \n",
       "SCYL3                       2.2144                   1.7912   \n",
       "C1orf112                    0.6608                   1.0363   \n",
       "\n",
       "          GTEX-RU1J-0426-SM-46MUK  GTEX-1212Z-0226-SM-59HLF       ...         \\\n",
       "TSPAN6                     2.3816                    4.5838       ...          \n",
       "TNMD                      -2.4659                   -2.9324       ...          \n",
       "DPM1                       5.1322                    4.0471       ...          \n",
       "SCYL3                      2.4753                    1.7912       ...          \n",
       "C1orf112                   1.3109                    0.1776       ...          \n",
       "\n",
       "          TCGA-AB-2965-03  TCGA-AB-2936-03  TCGA-AB-2839-03  TCGA-AB-2879-03  \\\n",
       "TSPAN6            -1.2828          -3.3076          -3.1714          -0.8084   \n",
       "TNMD              -9.9658          -9.9658          -9.9658          -9.9658   \n",
       "DPM1               5.2510           4.6566           4.7603           4.4156   \n",
       "SCYL3              2.6670           3.9147           3.3435           3.5535   \n",
       "C1orf112           3.1939           4.3896           3.5174           3.7105   \n",
       "\n",
       "          TCGA-AB-2886-03  TCGA-AB-2901-03  TCGA-AB-2862-03  TCGA-AB-2956-03  \\\n",
       "TSPAN6            -9.9658          -0.1828          -4.2934          -4.6082   \n",
       "TNMD              -9.9658          -9.9658          -9.9658          -9.9658   \n",
       "DPM1               4.8719           4.8625           4.5361           5.3345   \n",
       "SCYL3              3.5249           3.5060           2.1509           2.9929   \n",
       "C1orf112           4.8954           4.5681           2.6918           4.0304   \n",
       "\n",
       "          TCGA-AB-2987-03  TCGA-AB-2868-03  \n",
       "TSPAN6            -1.2481          -0.9132  \n",
       "TNMD              -3.4580          -9.9658  \n",
       "DPM1               5.1223           4.1285  \n",
       "SCYL3              2.5011           2.9507  \n",
       "C1orf112           2.4985           3.1491  \n",
       "\n",
       "[5 rows x 19260 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a new data frame replacing the ensembl based index with hugo dropping any where there is no conversion\n",
    "tcga_target_gtex_expression_hugo = tcga_target_gtex_expression.copy()\n",
    "tcga_target_gtex_expression_hugo.index = ensemble_to_hugo.reindex(tcga_target_gtex_expression.index).geneName.values\n",
    "tcga_target_gtex_expression_hugo = tcga_target_gtex_expression_hugo[tcga_target_gtex_expression_hugo.index.notnull()]\n",
    "tcga_target_gtex_expression_hugo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GTEX-146FH-1726-SM-5QGQ2    1.000001e+06\n",
       "GTEX-WZTO-2926-SM-3NM9I     9.999974e+05\n",
       "TCGA-AB-2965-03             9.999970e+05\n",
       "dtype: float64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# While we're at it let's verify that the sum of all expression levels for a sample in TPM space sums to 1 million\n",
    "tcga_target_gtex_expression_hugo[[\"GTEX-146FH-1726-SM-5QGQ2\", \"GTEX-WZTO-2926-SM-3NM9I\", \"TCGA-AB-2965-03\"]].apply(np.exp2).apply(lambda x: x - 0.001).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 35s, sys: 22.3 s, total: 1min 57s\n",
      "Wall time: 1min 48s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Multiple Ensemble genes map to the same Hugo name. Each of these values has been normalized via log2(TPM+0.001)\n",
    "# so we convert back into TPM, compute the mean, and re-normalize.\n",
    "tcga_target_gtex_expression_hugo_tpm = tcga_target_gtex_expression_hugo \\\n",
    "    .apply(np.exp2).subtract(0.001).groupby(level=0).aggregate(np.mean).add(0.001).apply(np.log2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-UTHO-1226-SM-3GAEE</th>\n",
       "      <th>GTEX-146FH-1726-SM-5QGQ2</th>\n",
       "      <th>GTEX-QDT8-0126-SM-48TZ1</th>\n",
       "      <th>GTEX-QCQG-1326-SM-48U24</th>\n",
       "      <th>GTEX-WZTO-2926-SM-3NM9I</th>\n",
       "      <th>GTEX-12WSB-0126-SM-59HJN</th>\n",
       "      <th>GTEX-11VI4-0626-SM-5EQLO</th>\n",
       "      <th>GTEX-T5JC-0526-SM-32PM7</th>\n",
       "      <th>GTEX-RU1J-0426-SM-46MUK</th>\n",
       "      <th>GTEX-1212Z-0226-SM-59HLF</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-AB-2965-03</th>\n",
       "      <th>TCGA-AB-2936-03</th>\n",
       "      <th>TCGA-AB-2839-03</th>\n",
       "      <th>TCGA-AB-2879-03</th>\n",
       "      <th>TCGA-AB-2886-03</th>\n",
       "      <th>TCGA-AB-2901-03</th>\n",
       "      <th>TCGA-AB-2862-03</th>\n",
       "      <th>TCGA-AB-2956-03</th>\n",
       "      <th>TCGA-AB-2987-03</th>\n",
       "      <th>TCGA-AB-2868-03</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5S_rRNA</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>0.531436</td>\n",
       "      <td>0.506253</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_8S_rRNA</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.158432</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.158432</td>\n",
       "      <td>-3.128688</td>\n",
       "      <td>-1.994226</td>\n",
       "      <td>-2.759535</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-2.914171</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.749798</td>\n",
       "      <td>-2.950835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>3.4700</td>\n",
       "      <td>3.2435</td>\n",
       "      <td>4.2921</td>\n",
       "      <td>3.5424</td>\n",
       "      <td>4.1852</td>\n",
       "      <td>3.562100</td>\n",
       "      <td>2.6395</td>\n",
       "      <td>2.3649</td>\n",
       "      <td>2.7029</td>\n",
       "      <td>11.1325</td>\n",
       "      <td>...</td>\n",
       "      <td>0.911500</td>\n",
       "      <td>2.881900</td>\n",
       "      <td>2.689500</td>\n",
       "      <td>1.454700</td>\n",
       "      <td>3.560800</td>\n",
       "      <td>2.664700</td>\n",
       "      <td>4.5367</td>\n",
       "      <td>1.7141</td>\n",
       "      <td>0.971600</td>\n",
       "      <td>2.409000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.0440</td>\n",
       "      <td>1.2992</td>\n",
       "      <td>2.0535</td>\n",
       "      <td>1.5165</td>\n",
       "      <td>3.3336</td>\n",
       "      <td>1.590200</td>\n",
       "      <td>1.0711</td>\n",
       "      <td>-0.2845</td>\n",
       "      <td>0.8961</td>\n",
       "      <td>1.4808</td>\n",
       "      <td>...</td>\n",
       "      <td>2.782600</td>\n",
       "      <td>3.445100</td>\n",
       "      <td>3.300300</td>\n",
       "      <td>2.673800</td>\n",
       "      <td>4.438400</td>\n",
       "      <td>3.673700</td>\n",
       "      <td>4.1392</td>\n",
       "      <td>2.3981</td>\n",
       "      <td>1.546500</td>\n",
       "      <td>3.415600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19260 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GTEX-UTHO-1226-SM-3GAEE  GTEX-146FH-1726-SM-5QGQ2  \\\n",
       "5S_rRNA                    -9.9658                   -9.9658   \n",
       "5_8S_rRNA                  -9.9658                   -9.9658   \n",
       "7SK                        -9.9658                   -9.9658   \n",
       "A1BG                        3.4700                    3.2435   \n",
       "A1BG-AS1                    0.0440                    1.2992   \n",
       "\n",
       "           GTEX-QDT8-0126-SM-48TZ1  GTEX-QCQG-1326-SM-48U24  \\\n",
       "5S_rRNA                    -9.9658                  -9.9658   \n",
       "5_8S_rRNA                  -9.9658                  -9.9658   \n",
       "7SK                        -9.9658                  -9.9658   \n",
       "A1BG                        4.2921                   3.5424   \n",
       "A1BG-AS1                    2.0535                   1.5165   \n",
       "\n",
       "           GTEX-WZTO-2926-SM-3NM9I  GTEX-12WSB-0126-SM-59HJN  \\\n",
       "5S_rRNA                    -9.9658                 -9.965800   \n",
       "5_8S_rRNA                  -9.9658                 -9.965800   \n",
       "7SK                        -9.9658                 -4.158432   \n",
       "A1BG                        4.1852                  3.562100   \n",
       "A1BG-AS1                    3.3336                  1.590200   \n",
       "\n",
       "           GTEX-11VI4-0626-SM-5EQLO  GTEX-T5JC-0526-SM-32PM7  \\\n",
       "5S_rRNA                     -9.9658                  -9.9658   \n",
       "5_8S_rRNA                   -9.9658                  -9.9658   \n",
       "7SK                         -9.9658                  -9.9658   \n",
       "A1BG                         2.6395                   2.3649   \n",
       "A1BG-AS1                     1.0711                  -0.2845   \n",
       "\n",
       "           GTEX-RU1J-0426-SM-46MUK  GTEX-1212Z-0226-SM-59HLF       ...         \\\n",
       "5S_rRNA                    -9.9658                   -9.9658       ...          \n",
       "5_8S_rRNA                  -9.9658                   -9.9658       ...          \n",
       "7SK                        -9.9658                   -9.9658       ...          \n",
       "A1BG                        2.7029                   11.1325       ...          \n",
       "A1BG-AS1                    0.8961                    1.4808       ...          \n",
       "\n",
       "           TCGA-AB-2965-03  TCGA-AB-2936-03  TCGA-AB-2839-03  TCGA-AB-2879-03  \\\n",
       "5S_rRNA          -9.965800        -9.965800        -9.965800         0.531436   \n",
       "5_8S_rRNA        -9.965800        -9.965800        -9.965800        -9.965800   \n",
       "7SK              -4.158432        -3.128688        -1.994226        -2.759535   \n",
       "A1BG              0.911500         2.881900         2.689500         1.454700   \n",
       "A1BG-AS1          2.782600         3.445100         3.300300         2.673800   \n",
       "\n",
       "           TCGA-AB-2886-03  TCGA-AB-2901-03  TCGA-AB-2862-03  TCGA-AB-2956-03  \\\n",
       "5S_rRNA           0.506253        -9.965800          -9.9658          -9.9658   \n",
       "5_8S_rRNA        -9.965800        -9.965800          -9.9658          -9.9658   \n",
       "7SK              -9.965800        -2.914171          -9.9658          -9.9658   \n",
       "A1BG              3.560800         2.664700           4.5367           1.7141   \n",
       "A1BG-AS1          4.438400         3.673700           4.1392           2.3981   \n",
       "\n",
       "           TCGA-AB-2987-03  TCGA-AB-2868-03  \n",
       "5S_rRNA          -9.965800        -9.965800  \n",
       "5_8S_rRNA        -9.965800        -9.965800  \n",
       "7SK              -3.749798        -2.950835  \n",
       "A1BG              0.971600         2.409000  \n",
       "A1BG-AS1          1.546500         3.415600  \n",
       "\n",
       "[5 rows x 19260 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_target_gtex_expression_hugo_tpm.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download Treehouse Expression\n",
    "\n",
    "The Treehouse public compendium is in Hugo log2(tpm+1). We need to download and convert into lot2(tpm+0.001) to match our TCGA+TARGET+GTEXt dataset above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "treehouse_expression: samples=58581 genes=11078\n",
      "CPU times: user 101 ms, sys: 1.14 s, total: 1.24 s\n",
      "Wall time: 1.24 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if not os.path.exists(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\"):\n",
    "    print(\"Downloading Treehouse Public Compendium\")\n",
    "    r = requests.get(\"https://treehouse.xenahubs.net/download/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\",\n",
    "                     stream=True)\n",
    "    r.raise_for_status()\n",
    "    with open(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\", \"wb\") as f:\n",
    "        for chunk in r.iter_content(chunk_size=32768):\n",
    "            f.write(chunk)\n",
    "\n",
    "if not os.path.exists(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\"):\n",
    "    print(\"Converting expression to dataframe and storing in hdf5 file\")\n",
    "    expression = pd.read_csv(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\", \n",
    "                             sep=\"\\t\", index_col=0).dropna(axis=\"index\").astype(np.float32).sort_index(axis=\"index\")\n",
    "    expression.to_hdf(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\", \"expression\", mode=\"w\", format=\"fixed\")\n",
    "\n",
    "treehouse_expression = pd.read_hdf(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\", \"expression\")\n",
    "print(\"treehouse_expression: samples={} genes={}\".format(*treehouse_expression.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check that we don't have any null/nan at this point\n",
    "assert not tcga_target_gtex_expression_hugo_tpm.isnull().values.any()\n",
    "assert not treehouse_expression.isnull().values.any()\n",
    "\n",
    "# Make sure they have identical hugo gene indexes\n",
    "assert np.array_equal(tcga_target_gtex_expression_hugo_tpm.index, treehouse_expression.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert into log2(tpm+0.001)\n",
    "treehouse_expression_hugo_tpm = treehouse_expression.apply(np.exp2).subtract(1.0).add(0.001).apply(np.log2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NOTE and REMINDER\n",
    "\n",
    "The current public Treehouse compendium was created by combining expression values that map to the same Hugo gene identify by calculating the mean of their log2(tpm+1) values. As a result those values will not match perfectly with the same samples in the TCGA+TARGET+GTEX dataset. The next public compendium from Treehouse will calculate mean in TPM space. Continue on here but later we need to come back and update this - or calculate things the right way for the TH and TR samples from the raw data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene with maximum delta: Metazoa_SRP 6.039602\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    58581.000000\n",
       "mean         0.001218\n",
       "std          0.060920\n",
       "min         -0.013058\n",
       "25%         -0.000107\n",
       "50%         -0.000016\n",
       "75%         -0.000016\n",
       "max          6.039602\n",
       "Name: TCGA-ZQ-A9CR-01, dtype: float64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check to verify the TCGA+TARGET samples in the Treehouse compendium match TPM wise with our conversions above\n",
    "sample_id = \"TCGA-ZQ-A9CR-01\"\n",
    "\n",
    "np.allclose(tcga_target_gtex_expression_hugo_tpm[sample_id], treehouse_expression_hugo_tpm[sample_id], 1, 1)\n",
    "\n",
    "argmax = (tcga_target_gtex_expression_hugo_tpm[sample_id] - treehouse_expression_hugo_tpm[sample_id]).values.argmax()\n",
    "gene = tcga_target_gtex_expression_hugo_tpm.index[argmax]\n",
    "print(\"Gene with maximum delta:\", gene,\n",
    "      tcga_target_gtex_expression_hugo_tpm[sample_id][gene] - treehouse_expression_hugo_tpm[sample_id][gene])\n",
    "\n",
    "(tcga_target_gtex_expression_hugo_tpm[sample_id] - treehouse_expression_hugo_tpm[sample_id]).describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download and Normalize Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the sample labels from Xena ie clinical/phenotype information on each sample\n",
    "if not os.path.exists(\"data/TcgaTargetGTEX_phenotype.txt.gz\"):\n",
    "    with open(\"data/TcgaTargetGTEX_phenotype.txt.gz\", \"wb\") as f:\n",
    "        f.write(requests.get(\"https://toil.xenahubs.net/download/TcgaTargetGTEX_phenotype.txt.gz\").content)\n",
    "\n",
    "tcga_target_gtex_labels = pd.read_table(\n",
    "    \"data/TcgaTargetGTEX_phenotype.txt.gz\", compression=\"gzip\", \n",
    "    header=0, names=[\"id\", \"category\", \"disease\", \"primary_site\", \"sample_type\", \"gender\", \"study\"],\n",
    "    sep=\"\\t\", encoding=\"ISO-8859-1\", index_col=0, dtype=\"str\").sort_index(axis=\"index\")\n",
    "\n",
    "\n",
    "# Compute and add a tumor/normal column - TCGA and TARGET have some normal samples, GTEX is all normal.\n",
    "tcga_target_gtex_labels[\"tumor_normal\"] = tcga_target_gtex_labels.apply(\n",
    "    lambda row: \"Normal\" if row[\"sample_type\"] in [\"Cell Line\", \"Normal Tissue\", \"Solid Tissue Normal\"]\n",
    "    else \"Tumor\", axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58581, 19260)\n",
      "(19131, 7)\n"
     ]
    }
   ],
   "source": [
    "# assert np.array_equal(tcga_target_gtex_expression_hugo_tpm.index, tcga_target_gtex_labels.index)\n",
    "print(tcga_target_gtex_expression_hugo_tpm.shape)\n",
    "print(tcga_target_gtex_labels.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-POYW-1226-SM-5LZWQ</th>\n",
       "      <td>Lung</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Male</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PARCHB-03</th>\n",
       "      <td>Acute Lymphoblastic Leukemia</td>\n",
       "      <td>Acute Lymphoblastic Leukemia</td>\n",
       "      <td>White blood cell</td>\n",
       "      <td>Primary Blood Derived Cancer - Peripheral Blood</td>\n",
       "      <td>Male</td>\n",
       "      <td>TARGET</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BF-A5ER-01</th>\n",
       "      <td>Skin Cutaneous Melanoma</td>\n",
       "      <td>Skin Cutaneous Melanoma</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Male</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-FY-A4B4-01</th>\n",
       "      <td>Thyroid Carcinoma</td>\n",
       "      <td>Thyroid Carcinoma</td>\n",
       "      <td>Thyroid Gland</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Female</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              category  \\\n",
       "id                                                       \n",
       "GTEX-1117F-0226-SM-5GZZ7        Adipose - Subcutaneous   \n",
       "GTEX-POYW-1226-SM-5LZWQ                           Lung   \n",
       "TARGET-10-PARCHB-03       Acute Lymphoblastic Leukemia   \n",
       "TCGA-BF-A5ER-01                Skin Cutaneous Melanoma   \n",
       "TCGA-FY-A4B4-01                      Thyroid Carcinoma   \n",
       "\n",
       "                                               disease      primary_site  \\\n",
       "id                                                                         \n",
       "GTEX-1117F-0226-SM-5GZZ7        Adipose - Subcutaneous    Adipose Tissue   \n",
       "GTEX-POYW-1226-SM-5LZWQ                           Lung              Lung   \n",
       "TARGET-10-PARCHB-03       Acute Lymphoblastic Leukemia  White blood cell   \n",
       "TCGA-BF-A5ER-01                Skin Cutaneous Melanoma              Skin   \n",
       "TCGA-FY-A4B4-01                      Thyroid Carcinoma     Thyroid Gland   \n",
       "\n",
       "                                                              sample_type  \\\n",
       "id                                                                          \n",
       "GTEX-1117F-0226-SM-5GZZ7                                    Normal Tissue   \n",
       "GTEX-POYW-1226-SM-5LZWQ                                     Normal Tissue   \n",
       "TARGET-10-PARCHB-03       Primary Blood Derived Cancer - Peripheral Blood   \n",
       "TCGA-BF-A5ER-01                                             Primary Tumor   \n",
       "TCGA-FY-A4B4-01                                             Primary Tumor   \n",
       "\n",
       "                          gender   study tumor_normal  \n",
       "id                                                     \n",
       "GTEX-1117F-0226-SM-5GZZ7  Female    GTEX       Normal  \n",
       "GTEX-POYW-1226-SM-5LZWQ     Male    GTEX       Normal  \n",
       "TARGET-10-PARCHB-03         Male  TARGET        Tumor  \n",
       "TCGA-BF-A5ER-01             Male    TCGA        Tumor  \n",
       "TCGA-FY-A4B4-01           Female    TCGA        Tumor  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_target_gtex_labels[0:20000:4000].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>19130</td>\n",
       "      <td>19130</td>\n",
       "      <td>19126</td>\n",
       "      <td>19131</td>\n",
       "      <td>18972</td>\n",
       "      <td>19131</td>\n",
       "      <td>19131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>46</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Male</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1212</td>\n",
       "      <td>1212</td>\n",
       "      <td>1846</td>\n",
       "      <td>9185</td>\n",
       "      <td>10456</td>\n",
       "      <td>10535</td>\n",
       "      <td>10531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         category                    disease primary_site  \\\n",
       "count                       19130                      19130        19126   \n",
       "unique                         93                         93           46   \n",
       "top     Breast Invasive Carcinoma  Breast Invasive Carcinoma        Brain   \n",
       "freq                         1212                       1212         1846   \n",
       "\n",
       "          sample_type gender  study tumor_normal  \n",
       "count           19131  18972  19131        19131  \n",
       "unique             17      2      3            2  \n",
       "top     Primary Tumor   Male   TCGA        Tumor  \n",
       "freq             9185  10456  10535        10531  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_target_gtex_labels.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the tissue location as the class label for the purposes of stratification\n",
    "class_attribute = \"primary_site\"\n",
    "\n",
    "# Tumor vs. Normal is the binary attribute we'll use to train on\n",
    "label_attribute = \"tumor_normal\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort, transpose to machine learning standard of rows as samples and write out to hdf5 for fast loading into training notebooks\n",
    "tcga_target_gtex_expression_hugo_tpm_sorted = tcga_target_gtex_expression_hugo_tpm.sort_index(axis=\"index\").sort_index(axis=\"columns\")\n",
    "tcga_target_gtex_expression_hugo_tpm_sorted.to_hdf(\"data/tcga_target_gtex_expression_hugo_tpm.hd5\", \"expression\", mode=\"w\", format=\"fixed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort, transpose to machine learning standard of rows as samples and write out to hdf5 for fast loading into training notebooks\n",
    "treehouse_expression_hugo_tpm_sorted = treehouse_expression_hugo_tpm.sort_index(axis=\"index\").sort_index(axis=\"columns\")\n",
    "treehouse_expression_hugo_tpm.to_hdf(\"data/treehouse_expression_hugo_tpm.hd5\", \"expression\", mode=\"w\", format=\"fixed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove rows where the class is null or the sample is missing\n",
    "Y_not_null = Y[pd.notnull(Y[class_attribute])]\n",
    "intersection = X.index.intersection(Y_not_null.index)\n",
    "X_clean = X[X.index.isin(intersection)]\n",
    "Y_clean = Y[Y.index.isin(intersection)]\n",
    "\n",
    "# Make sure the label and example samples are in the same order\n",
    "assert(X_clean.index.equals(Y_clean.index))\n",
    "\n",
    "print(intersection.shape[0], \"samples with non-null labels\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert tumor/normal labels to binary 1/0\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "encoder = LabelEncoder()\n",
    "y_binary = encoder.fit_transform(Y_clean[\"tumor_normal\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert classes into numbers\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "encoder = LabelEncoder()\n",
    "encoder.fit(Y_clean[class_attribute].values)\n",
    "classes = encoder.transform(Y_clean[class_attribute])\n",
    "print(\"Total classes for stratification:\", len(set(classes)))\n",
    "class_labels = encoder.classes_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# Split into stratified training and test sets based on classes (i.e. tissue type) so that we have equal\n",
    "# proportions of each tissue type in the train and test sets\n",
    "from sklearn.model_selection import StratifiedShuffleSplit\n",
    "split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)\n",
    "for train_index, test_index in split.split(X_clean.values, Y_clean[class_attribute]):\n",
    "    X_train, X_test = X.values[train_index], X_clean.values[test_index]\n",
    "    y_train, y_test = y_binary[train_index], y_binary[test_index]\n",
    "    classes_train, classes_test = classes[train_index], classes[test_index]\n",
    "    sample_labels_train, sample_labels_test = X.index[train_index], X.index[test_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Write to an h5 file for training (see above for details on each dataset)\n",
    "\"\"\"\n",
    "with h5py.File(\"data/tumor_normal.h5\", \"w\") as f:\n",
    "    f.create_dataset('X_train', X_train.shape, dtype='f')[:] = X_train\n",
    "    f.create_dataset('X_test', X_test.shape, dtype='f')[:] = X_test\n",
    "    f.create_dataset('y_train', y_train.shape, dtype='i')[:] = y_train\n",
    "    f.create_dataset('y_test', y_test.shape, dtype='i')[:] = y_test\n",
    "    f.create_dataset('classes_train', y_train.shape, dtype='i')[:] = classes_train\n",
    "    f.create_dataset('classes_test', y_test.shape, dtype='i')[:] = classes_test\n",
    "    f.create_dataset('features', X_clean.columns.shape, 'S10', \n",
    "                     [l.encode(\"ascii\", \"ignore\") for l in X_clean.columns.values])\n",
    "    f.create_dataset('labels', (2, 1), 'S10', \n",
    "                     [l.encode(\"ascii\", \"ignore\") for l in [\"Normal\", \"Tumor\"]])\n",
    "    f.create_dataset('class_labels', (len(class_labels), 1), 'S10', \n",
    "                     [l.encode(\"ascii\", \"ignore\") for l in class_labels])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as pyplot\n",
    "pyplot.hist(classes_train, alpha=0.5, label='Train')\n",
    "pyplot.hist(classes_test, alpha=0.5, label='Test')\n",
    "pyplot.legend(loc='upper right')\n",
    "pyplot.title(\"Class (Primary Site) distribution between train and test\")\n",
    "pyplot.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
